### ISSUE TITLE
New output format: Markdown

### ISSUE THREAD

**Issue by dirkrombauts** [picklesdoc/pickles#237]
Use Markdown as an additional output format, to complement the set of Html, Dhtml, Word, Excel and CucumberJson.

**Comment by dstj** [comment]
If you take a look at [this commit](https://github.com/mediaclip/pickles/commit/08567d4cc9fc2a3729db118a97c74f0c0fe28ebc) on our fork's master branch, you'll see that we used Markdownify as Markdown parsing engine. We found that the output was better...

I didn't want want to include all of that in the pull request #252 I created, but I think allowing multiple Markdown parsers would be a good improvement.

**Comment by dirkrombauts** [comment]
Are we perhaps talking about different things here? The markdown parser does not have any bearing on the output formats.

**Comment by dirkrombauts** [comment]
Oh, I get it - the markdown parser is also the markdown renderer.

**Comment by dstj** [comment]
Actually, I made a mistake, I read "Mardown parser" instead of "output format". I saw Markdown and got excited I should have created a new issue post...

**Comment by allustin** [comment]
:+1: for this feature

@dirkrombauts i use a some hack for this right now

```
pickles.exe -f ./features -l ru -o .\docs\ -df word --sn "My Super Project" 
pandoc -f docx ".\docs\My Super Project.docx" -t markdown_github >> ./FEATURES.md
```

but this is only a hack

**Comment by darrencomeau** [comment]
This would be really useful when committing the results back to source control as the likes of GitHub, Azure DevOps, and BitBucket don't render HTML but would display your Markdown content faithfully.

You could keep your test results along side the code, keeping them immutably linked.

Is anyone actively working on this?

**Comment by dirkrombauts** [comment]
Hi,
As far as I know, no-one is working on this at the moment. So feel free to have a go at this yourself 🙂

**Comment by darrencomeau** [comment]
**I'd certainly like to give it a go.** 🙂

I'll start putting up some suggested formats for discussion and over the next few weeks I'll see what I can do!

**Comment by darrencomeau** [comment]
Just starting the ball rolling on ideas for the formatting.

Has anyone got any suggestions for what they'd like to see in a markdown document format?

- Single page, multipage
- Links across headers
- Summary of execution
- Any images/logos
- etc...


# Features

---

Generated on: 14 October 2018 11:34:52
2 tests : 1 passing : 1 failing

---

## My Project/Assembly/Class?

---

### *`@MyTag`*
### My First Feature

In order to achieve something as a person I want something else to happen

> #### Background
> **Given** I do something
> **And** I do something else



> #### <img src="https://centredmv.com/wp-content/uploads/2017/01/Error.png" alt="fail" width="12" height="12" /> A scenario where something happens
> **When** I do something
> **Then** this result happens


>*`@Tag`* *`@Other`*
>#### <img src="https://centredmv.com/wp-content/uploads/2017/01/Correct.png" alt="pass" width="12" height="12" /> A scenario where something else happens
> **When** I do something
> **Then** this result happens

---

###  *`@MyTag`*
### My Second Feature

In order to achieve something as a person I want something else to happen

> #### <img src="https://centredmv.com/wp-content/uploads/2017/01/Error.png" alt="fail" width="12" height="12" /> A scenario where something happens
> **When** I do something
> **Then** this result happens


>*`@Ignored`*
>#### <img src="http://www.free-icons-download.net/images/orange-question-mark-icon-24513.png" alt="inconclusive" width="12" height="12" /> A scenario where something else happens
> **When** I do something
> **Then** this result happens

**Comment by darrencomeau** [comment]
I'm starting to implement this output format now...

Source code is in my repo on the [markdown_output](https://github.com/VulcanAnalytics/pickles/tree/feature/markdown_output) branch. If there's anything anyone wants contribute then I can look at pull requests to that branch.

There is a [project board](https://github.com/VulcanAnalytics/pickles/projects/1) where you can track progress.


If there are any features that you want to have included in this output format please [raise an issue](https://github.com/VulcanAnalytics/pickles/issues)


The target format has moved to the wiki page [Concept-Markdown-Output](https://github.com/VulcanAnalytics/pickles/wiki/Concept-Markdown-Output)
while current progress can quickly be seen at [Current-Markdown-Output](https://github.com/VulcanAnalytics/pickles/wiki/Current-Markdown-Output)

Cheers,

Darren.

**Comment by dirkrombauts** [comment]
Cool - my implementation time is limited, but if you have any technical questions, I'll do my best to answer them!

**Comment by darrencomeau** [comment]
Hi @dirkrombauts, how would you like the pull requests? A single request once all changes are complete or incremental so that you can keep tabs on the implementation and release when you feel there's enough functionality?

I've got the title

> # Features

complete, not ground breaking but marks an understanding of the plumbing etc.

**Comment by dirkrombauts** [comment]
If you ask me: create a pull request now, and put something like "Work in Progress" in the title. Whenever you add commits to your branch, the pull request gets updated. That way I can have a look at it from time to time.

**Comment by darrencomeau** [comment]
Great, I'll go ahead and do that.

**Comment by dirkrombauts** [comment]
That looks very thorough so far!

**Comment by darrencomeau** [comment]
Thanks.

There's a build issue which I'm ignoring atm, do you know anything which might fix it? Looks to be all related to the cucumber code;

> Errors, Failures and Warnings
> 4300
> 4301
> 1) TearDown Error : PicklesDoc.Pickles.DocumentationBuilders.Cucumber.UnitTests.FormattingAFeatureFeature
> 4302
> System.IO.FileLoadException : Could not load file or assembly 'TechTalk.SpecFlow, Version=2.3.2.0, Culture=neutral, PublicKeyToken=0778194805d6db41' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
> 4303
> TearDown : System.IO.FileLoadException : Could not load file or assembly 'TechTalk.SpecFlow, Version=2.3.2.0, Culture=neutral, PublicKeyToken=0778194805d6db41' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
> 4304
>    at PicklesDoc.Pickles.DocumentationBuilders.Cucumber.UnitTests.FormattingAFeatureFeature.FeatureSetup()
> 4305
> --TearDown
> 4306
>    at PicklesDoc.Pickles.DocumentationBuilders.Cucumber.UnitTests.FormattingAFeatureFeature.FeatureTearDown()
> 4307
> 4308
> 2) Error : PicklesDoc.Pickles.DocumentationBuilders.Cucumber.UnitTests.FormattingAFeatureFeature.AFeatureWithATable
> 4309
> OneTimeSetUp: System.IO.FileLoadException : Could not load file or assembly 'TechTalk.SpecFlow, Version=2.3.2.0, Culture=neutral, PublicKeyToken=0778194805d6db41' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
> 4310
> 4311
> 3) Error : PicklesDoc.Pickles.DocumentationBuilders.Cucumber.UnitTests.FormattingAFeatureFeature.ASimpleFeature
> 4312
> OneTimeSetUp: System.IO.FileLoadException : Could not load file or assembly 'TechTalk.SpecFlow, Version=2.3.2.0, Culture=neutral, PublicKeyToken=0778194805d6db41' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)

**Comment by dirkrombauts** [comment]
Do all projects use the same version of the SpecFlow package? Does the Cucumber.UnitTests project have an explicit reference to the SpecFlow package?

**Comment by darrencomeau** [comment]
Hi @dirkrombauts, the pull request now has most features to make a usable document and is at a stage which I would consider ready for beta testing.

While there are still some things which can be added to the feature and likely some bugs to discover; what is the next step and how should it proceed?

Is there anything which you would want to see included still or altered?

Thanks

**Comment by patriml** [comment]
I'm interested in this feature, can I help somehow?

**Comment by dirkrombauts** [comment]
Yes, you can! There‘s a pull request for a markdown output format. It would help if you could put that version to the test. Use it on your features and see if the output matches your expectations.

**Comment by dirkrombauts** [comment]
Released in version [v2.21.0](https://github.com/picklesdoc/pickles/releases/tag/v2.21.0).

**Comment by dirkrombauts** [comment]
It's been an amazing time for me to work on Pickles. Now it's finally time for me to lay down the mantle and move on. I am leaving Pickles completely.

I am closing this issue, so that the next maintainer of this repository can start from a clean slate.

Do you want to take over active development and maintenance at Pickles? Contact me directly at dirk dot rombauts at picklesdoc dot com. I will hand over everything Pickles-related to you. This email address will remain active until 11 December 2020.
